File: cplusplus/astronauts-in-space.cpp
//  To parse this JSON data, first install
//
//      Boost     http://www.boost.org
//      json.hpp  https://github.com/nlohmann/json
//
//  Then include this file, and then do
//
//     AstronautsInSpace data = nlohmann::json::parse(jsonString);
//     IssCurrentLocation data = nlohmann::json::parse(jsonString);

#pragma once

#include "json.hpp"

#include <boost/optional.hpp>
#include <stdexcept>
#include <regex>
namespace quicktype {
    using nlohmann::json;

    inline json get_untyped(const json &j, const char *property) {
        if (j.find(property) != j.end()) {
            return j.at(property).get<json>();
        }
        return json();
    }

    class Person {
        public:
        Person() = default;
        virtual ~Person() = default;

        private:
        std::string craft;
        std::string name;

        public:
        const std::string & get_craft() const { return craft; }
        std::string & get_mutable_craft() { return craft; }
        void set_craft(const std::string& value) { this->craft = value; }

        const std::string & get_name() const { return name; }
        std::string & get_mutable_name() { return name; }
        void set_name(const std::string& value) { this->name = value; }
    };

    class AstronautsInSpace {
        public:
        AstronautsInSpace() = default;
        virtual ~AstronautsInSpace() = default;

        private:
        std::string message;
        std::vector<Person> people;
        int64_t number;

        public:
        const std::string & get_message() const { return message; }
        std::string & get_mutable_message() { return message; }
        void set_message(const std::string& value) { this->message = value; }

        const std::vector<Person> & get_people() const { return people; }
        std::vector<Person> & get_mutable_people() { return people; }
        void set_people(const std::vector<Person>& value) { this->people = value; }

        const int64_t & get_number() const { return number; }
        int64_t & get_mutable_number() { return number; }
        void set_number(const int64_t& value) { this->number = value; }
    };

    class IssPosition {
        public:
        IssPosition() = default;
        virtual ~IssPosition() = default;

        private:
        std::string longitude;
        std::string latitude;

        public:
        const std::string & get_longitude() const { return longitude; }
        std::string & get_mutable_longitude() { return longitude; }
        void set_longitude(const std::string& value) { this->longitude = value; }

        const std::string & get_latitude() const { return latitude; }
        std::string & get_mutable_latitude() { return latitude; }
        void set_latitude(const std::string& value) { this->latitude = value; }
    };

    class IssCurrentLocation {
        public:
        IssCurrentLocation() = default;
        virtual ~IssCurrentLocation() = default;

        private:
        IssPosition iss_position;
        int64_t timestamp;
        std::string message;

        public:
        const IssPosition & get_iss_position() const { return iss_position; }
        IssPosition & get_mutable_iss_position() { return iss_position; }
        void set_iss_position(const IssPosition& value) { this->iss_position = value; }

        const int64_t & get_timestamp() const { return timestamp; }
        int64_t & get_mutable_timestamp() { return timestamp; }
        void set_timestamp(const int64_t& value) { this->timestamp = value; }

        const std::string & get_message() const { return message; }
        std::string & get_mutable_message() { return message; }
        void set_message(const std::string& value) { this->message = value; }
    };
}

namespace nlohmann {
    inline void from_json(const json& _j, quicktype::Person& _x) {
        _x.set_craft( _j.at("craft").get<std::string>() );
        _x.set_name( _j.at("name").get<std::string>() );
    }

    inline void to_json(json& _j, const quicktype::Person& _x) {
        _j = json::object();
        _j["craft"] = _x.get_craft();
        _j["name"] = _x.get_name();
    }

    inline void from_json(const json& _j, quicktype::AstronautsInSpace& _x) {
        _x.set_message( _j.at("message").get<std::string>() );
        _x.set_people( _j.at("people").get<std::vector<quicktype::Person>>() );
        _x.set_number( _j.at("number").get<int64_t>() );
    }

    inline void to_json(json& _j, const quicktype::AstronautsInSpace& _x) {
        _j = json::object();
        _j["message"] = _x.get_message();
        _j["people"] = _x.get_people();
        _j["number"] = _x.get_number();
    }

    inline void from_json(const json& _j, quicktype::IssPosition& _x) {
        _x.set_longitude( _j.at("longitude").get<std::string>() );
        _x.set_latitude( _j.at("latitude").get<std::string>() );
    }

    inline void to_json(json& _j, const quicktype::IssPosition& _x) {
        _j = json::object();
        _j["longitude"] = _x.get_longitude();
        _j["latitude"] = _x.get_latitude();
    }

    inline void from_json(const json& _j, quicktype::IssCurrentLocation& _x) {
        _x.set_iss_position( _j.at("iss_position").get<quicktype::IssPosition>() );
        _x.set_timestamp( _j.at("timestamp").get<int64_t>() );
        _x.set_message( _j.at("message").get<std::string>() );
    }

    inline void to_json(json& _j, const quicktype::IssCurrentLocation& _x) {
        _j = json::object();
        _j["iss_position"] = _x.get_iss_position();
        _j["timestamp"] = _x.get_timestamp();
        _j["message"] = _x.get_message();
    }
}

File: python/astronauts-in-space.py
# To use this code, make sure you
#
#     import json
#
# and then, to convert JSON from a string, do
#
#     result = astronauts_in_space_from_dict(json.loads(json_string))
#     result = iss_current_location_from_dict(json.loads(json_string))

from typing import Any, List, TypeVar, Callable, Type, cast


T = TypeVar("T")


def from_str(x: Any) -> str:
    assert isinstance(x, str)
    return x


def from_list(f: Callable[[Any], T], x: Any) -> List[T]:
    assert isinstance(x, list)
    return [f(y) for y in x]


def from_int(x: Any) -> int:
    assert isinstance(x, int) and not isinstance(x, bool)
    return x


def to_class(c: Type[T], x: Any) -> dict:
    assert isinstance(x, c)
    return cast(Any, x).to_dict()


class Person:
    craft: str
    name: str

    def __init__(self, craft: str, name: str) -> None:
        self.craft = craft
        self.name = name

    @staticmethod
    def from_dict(obj: Any) -> 'Person':
        assert isinstance(obj, dict)
        craft = from_str(obj.get("craft"))
        name = from_str(obj.get("name"))
        return Person(craft, name)

    def to_dict(self) -> dict:
        result: dict = {}
        result["craft"] = from_str(self.craft)
        result["name"] = from_str(self.name)
        return result


class AstronautsInSpace:
    message: str
    people: List[Person]
    number: int

    def __init__(self, message: str, people: List[Person], number: int) -> None:
        self.message = message
        self.people = people
        self.number = number

    @staticmethod
    def from_dict(obj: Any) -> 'AstronautsInSpace':
        assert isinstance(obj, dict)
        message = from_str(obj.get("message"))
        people = from_list(Person.from_dict, obj.get("people"))
        number = from_int(obj.get("number"))
        return AstronautsInSpace(message, people, number)

    def to_dict(self) -> dict:
        result: dict = {}
        result["message"] = from_str(self.message)
        result["people"] = from_list(lambda x: to_class(Person, x), self.people)
        result["number"] = from_int(self.number)
        return result


class IssPosition:
    longitude: str
    latitude: str

    def __init__(self, longitude: str, latitude: str) -> None:
        self.longitude = longitude
        self.latitude = latitude

    @staticmethod
    def from_dict(obj: Any) -> 'IssPosition':
        assert isinstance(obj, dict)
        longitude = from_str(obj.get("longitude"))
        latitude = from_str(obj.get("latitude"))
        return IssPosition(longitude, latitude)

    def to_dict(self) -> dict:
        result: dict = {}
        result["longitude"] = from_str(self.longitude)
        result["latitude"] = from_str(self.latitude)
        return result


class ISSCurrentLocation:
    iss_position: IssPosition
    timestamp: int
    message: str

    def __init__(self, iss_position: IssPosition, timestamp: int, message: str) -> None:
        self.iss_position = iss_position
        self.timestamp = timestamp
        self.message = message

    @staticmethod
    def from_dict(obj: Any) -> 'ISSCurrentLocation':
        assert isinstance(obj, dict)
        iss_position = IssPosition.from_dict(obj.get("iss_position"))
        timestamp = from_int(obj.get("timestamp"))
        message = from_str(obj.get("message"))
        return ISSCurrentLocation(iss_position, timestamp, message)

    def to_dict(self) -> dict:
        result: dict = {}
        result["iss_position"] = to_class(IssPosition, self.iss_position)
        result["timestamp"] = from_int(self.timestamp)
        result["message"] = from_str(self.message)
        return result


def astronauts_in_space_from_dict(s: Any) -> AstronautsInSpace:
    return AstronautsInSpace.from_dict(s)


def astronauts_in_space_to_dict(x: AstronautsInSpace) -> Any:
    return to_class(AstronautsInSpace, x)


def iss_current_location_from_dict(s: Any) -> ISSCurrentLocation:
    return ISSCurrentLocation.from_dict(s)


def iss_current_location_to_dict(x: ISSCurrentLocation) -> Any:
    return to_class(ISSCurrentLocation, x)
